src/tensors/diagonal.jl

```diff
@@ -52,6 +52,14 @@ function DiagonalTensorMap(data::DenseVector{T}, V::IndexSpace) where {T}
     return DiagonalTensorMap{T}(data, V)
 end
 
+function DiagonalTensorMap(t::AbstractTensorMap{T,S,1,1}) where {T,S}
+    isa(t, DiagonalTensorMap) && return t
+    @assert domain(t) == codomain(t) "Domain and codomain of the input tensor are different."
+    @assert all(Diagonal(b) == b for (k, b) in blocks(t)) "Input tensor is not diagonal."
```

Comment:
I'd prefer somethign like: 

```julia
domain(t) == codomain(t) || throw(SpaceMismatch("DiagonalTensorMap requires equal domain and codomain"))
all(LinearAlgebra.isdiag ∘ last, blocks(t)) || throw(ArgumentError("DiagonalTensorMap requires input tensor that is diagonal"))
```

Reply:
Also, it seems like `LinearAlgebra.Diagonal(m::Matrix)` happily discards the off-diagonal data and just takes the diagonal of `m`. Should we follow this behaviour or not? If we do, then I the check for `isdiag` should move to `convert`, otherwise it is indeed implemented here.

Reply:
Yes, this is actually how it is supposed to be: a converter is lossless, and should work both ways, (since this might happen silently when assigning fields/entries of typed containers), but a constructor does not require this property. https://docs.julialang.org/en/v1/manual/conversion-and-promotion/#Conversion-vs.-Construction